pkg/cache/cache.go

```diff
@@ -87,47 +98,72 @@ func init() {
+	return c.GetOrNotify(name, namespace, nil)
+}
+
+// GetOrNotify will return the cached configuration of the given ServiceAccount.
+// It will first look at the set of ServiceAccounts configured using annotations. If none is found, it will register
+// handler to be notified as soon as a ServiceAccount with given key is populated to the cache. Afterwards it will check
+// for a ServiceAccount configured through the pod-identity-webhook ConfigMap.
+func (c *serviceAccountCache) GetOrNotify(name, namespace string, handler chan any) Response {
+	result := Response{
+		TokenExpiration: pkg.DefaultTokenExpiration,
+	}
 	klog.V(5).Infof("Fetching sa %s/%s from cache", namespace, name)
 	{
-		resp := c.getSA(name, namespace)
+		resp := c.getSAorNotify(name, namespace, handler)
+		if resp != nil {
+			result.FoundInSACache = true
+		}
 		if resp != nil && resp.RoleARN != "" {
-			return resp.RoleARN, resp.Audience, resp.UseRegionalSTS, resp.TokenExpiration
+			result.RoleARN = resp.RoleARN
+			result.Audience = resp.Audience
+			result.UseRegionalSTS = resp.UseRegionalSTS
+			result.TokenExpiration = resp.TokenExpiration
+			return result
 		}
 	}
 	{
 		resp := c.getCM(name, namespace)
 		if resp != nil {
-			return resp.RoleARN, resp.Audience, resp.UseRegionalSTS, resp.TokenExpiration
+			result.FoundInCMCache = true
+			result.RoleARN = resp.RoleARN
+			result.Audience = resp.Audience
+			result.UseRegionalSTS = resp.UseRegionalSTS
+			result.TokenExpiration = resp.TokenExpiration
+			return result
 		}
 	}
 	klog.V(5).Infof("Service account %s/%s not found in cache", namespace, name)
-	return "", "", false, pkg.DefaultTokenExpiration
+	return result
 }
 
 // GetCommonConfigurations returns the common configurations that also applies to the new mutation method(i.e Container Credentials).
 // The config file for the container credentials does not contain "TokenExpiration" or "UseRegionalSTS". For backward compatibility,
 // Use these fields if they are set in the sa annotations or config map.
 func (c *serviceAccountCache) GetCommonConfigurations(name, namespace string) (useRegionalSTS bool, tokenExpiration int64) {
-	if resp := c.getSA(name, namespace); resp != nil {
+	if resp := c.getSAorNotify(name, namespace, nil); resp != nil {
 		return resp.UseRegionalSTS, resp.TokenExpiration
 	} else if resp := c.getCM(name, namespace); resp != nil {
 		return resp.UseRegionalSTS, resp.TokenExpiration
 	}
 	return false, pkg.DefaultTokenExpiration
 }
 
-func (c *serviceAccountCache) getSA(name, namespace string) *CacheResponse {
+func (c *serviceAccountCache) getSAorNotify(name, namespace string, handler chan any) *Entry {
 	c.mu.RLock()
 	defer c.mu.RUnlock()
 	resp, ok := c.saCache[namespace+"/"+name]
-	if !ok {
+	if !ok && handler != nil {
+		klog.V(5).Infof("Service Account %s/%s not found in cache, adding notification handler", namespace, name)
+		c.notificationHandlers[namespace+"/"+name] = handler
```

Comment:
100% true and I think it is not only a performace bug, it can also be a functional bug. But actually I don't like the idea of retrieving the channel too much. Do you think this is working that multiple goroutines are notified throug a single notification in a single channel? I never tested this tbh. Furthermore, from my point of view it would be far more expressive and clear when we replace `notificationHandlers   map[string]chan struct{}` with `notificationHandlers   map[string][]chan struct{}` or even better with a slice of structs containing also the computed timeout and let notification only happen when channel is not closed and it is not after timeout. 

Happy to get to know your thoughts about that.

Reply:
I now implemented like you proposed with a slightly different interface:

There's only one Get() method remaining. It receives a struct with all former parameters and a switch indicating if the caller want's to be notified or not. If notification is required, a channel is delivered to the caller through the `cache.Response` struct. I think that's clean and it s going to work with simulataneous hook requests dealing with the same SA.